File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/src/Config/GreyOptions.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Config;

/**
 * Configuration options for the Grey SDK client.
 */
final class GreyOptions
{
    public const DEFAULT_TIMEOUT_MS = 30000;
    public const DEFAULT_PORT = 443;

    public function __construct(
        private readonly string $host,
        private readonly int $port = self::DEFAULT_PORT,
        private readonly bool $useTls = true,
        private readonly int $timeoutMs = self::DEFAULT_TIMEOUT_MS,
        private readonly array $metadata = [],
    ) {}

    /**
     * Gets the server host.
     */
    public function getHost(): string
    {
        return $this->host;
    }

    /**
     * Gets the server port.
     */
    public function getPort(): int
    {
        return $this->port;
    }

    /**
     * Returns true if TLS should be used.
     */
    public function useTls(): bool
    {
        return $this->useTls;
    }

    /**
     * Gets the request timeout in milliseconds.
     */
    public function getTimeoutMs(): int
    {
        return $this->timeoutMs;
    }

    /**
     * Gets the timeout in seconds for Guzzle.
     */
    public function getTimeoutSeconds(): float
    {
        return $this->timeoutMs / 1000.0;
    }

    /**
     * Gets additional metadata headers.
     */
    public function getMetadata(): array
    {
        return $this->metadata;
    }

    /**
     * Gets the base URL for HTTP requests.
     */
    public function getBaseUrl(): string
    {
        $scheme = $this->useTls ? 'https' : 'http';
        return "{$scheme}://{$this->host}:{$this->port}";
    }

    /**
     * Creates options for local development.
     */
    public static function local(int $port = 8080): self
    {
        return new self(
            host: 'localhost',
            port: $port,
            useTls: false,
            timeoutMs: self::DEFAULT_TIMEOUT_MS,
        );
    }

    /**
     * Creates options for production with TLS.
     */
    public static function production(string $host, int $port = 443): self
    {
        return new self(
            host: $host,
            port: $port,
            useTls: true,
            timeoutMs: self::DEFAULT_TIMEOUT_MS,
        );
    }

    /**
     * Returns new options with a different timeout.
     */
    public function withTimeout(int $timeoutMs): self
    {
        return new self(
            $this->host,
            $this->port,
            $this->useTls,
            $timeoutMs,
            $this->metadata,
        );
    }

    /**
     * Returns new options with additional metadata merged.
     */
    public function withMetadata(array $metadata): self
    {
        return new self(
            $this->host,
            $this->port,
            $this->useTls,
            $this->timeoutMs,
            array_merge($this->metadata, $metadata),
        );
    }

    /**
     * Converts options to an array.
     */
    public function toArray(): array
    {
        return [
            'host' => $this->host,
            'port' => $this->port,
            'useTls' => $this->useTls,
            'timeoutMs' => $this->timeoutMs,
            'metadata' => $this->metadata,
        ];
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-erlang/src/grey_options.erl
%%%-------------------------------------------------------------------
%%% @doc Grey SDK configuration options module.
%%%
%%% Provides configuration options for the Grey SDK client.
%%% @end
%%%-------------------------------------------------------------------
-module(grey_options).

-include("grey_sdk.hrl").

%% API exports
-export([
    new/3,
    new/4,
    new/5,
    local/0,
    local/1,
    production/1,
    host/1,
    port/1,
    use_tls/1,
    timeout_ms/1,
    metadata/1,
    endpoint/1,
    to_map/1
]).

%%====================================================================
%% API functions
%%====================================================================

%% @doc Creates new options with required parameters.
-spec new(binary(), pos_integer(), boolean()) -> #grey_options{}.
new(Host, Port, UseTls) ->
    new(Host, Port, UseTls, 30000).

%% @doc Creates new options with timeout.
-spec new(binary(), pos_integer(), boolean(), pos_integer()) -> #grey_options{}.
new(Host, Port, UseTls, TimeoutMs) ->
    new(Host, Port, UseTls, TimeoutMs, #{}).

%% @doc Creates new options with all parameters.
-spec new(binary(), pos_integer(), boolean(), pos_integer(), map()) -> #grey_options{}.
new(Host, Port, UseTls, TimeoutMs, Metadata) ->
    validate_host(Host),
    validate_port(Port),
    validate_timeout(TimeoutMs),
    #grey_options{
        host = Host,
        port = Port,
        use_tls = UseTls,
        timeout_ms = TimeoutMs,
        metadata = Metadata
    }.

%% @doc Creates options for local development.
-spec local() -> #grey_options{}.
local() ->
    local(50051).

-spec local(pos_integer()) -> #grey_options{}.
local(Port) ->
    new(<<"localhost">>, Port, false).

%% @doc Creates options for production.
-spec production(binary()) -> #grey_options{}.
production(Host) ->
    new(Host, 443, true).

%% @doc Gets the host.
-spec host(#grey_options{}) -> binary().
host(#grey_options{host = Host}) ->
    Host.

%% @doc Gets the port.
-spec port(#grey_options{}) -> pos_integer().
port(#grey_options{port = Port}) ->
    Port.

%% @doc Gets the TLS setting.
-spec use_tls(#grey_options{}) -> boolean().
use_tls(#grey_options{use_tls = UseTls}) ->
    UseTls.

%% @doc Gets the timeout in milliseconds.
-spec timeout_ms(#grey_options{}) -> pos_integer().
timeout_ms(#grey_options{timeout_ms = TimeoutMs}) ->
    TimeoutMs.

%% @doc Gets the metadata.
-spec metadata(#grey_options{}) -> map().
metadata(#grey_options{metadata = Metadata}) ->
    Metadata.

%% @doc Gets the full endpoint string.
-spec endpoint(#grey_options{}) -> binary().
endpoint(#grey_options{host = Host, port = Port}) ->
    list_to_binary(io_lib:format("~s:~p", [Host, Port])).

%% @doc Converts options to a map.
-spec to_map(#grey_options{}) -> map().
to_map(#grey_options{host = Host, port = Port, use_tls = UseTls,
                     timeout_ms = TimeoutMs, metadata = Metadata}) ->
    #{
        host => Host,
        port => Port,
        use_tls => UseTls,
        timeout_ms => TimeoutMs,
        metadata => Metadata
    }.

%%====================================================================
%% Internal functions
%%====================================================================

validate_host(Host) when is_binary(Host), byte_size(Host) > 0 ->
    ok;
validate_host(_) ->
    error(badarg).

validate_port(Port) when is_integer(Port), Port > 0, Port =< 65535 ->
    ok;
validate_port(_) ->
    error(badarg).

validate_timeout(Timeout) when is_integer(Timeout), Timeout > 0 ->
    ok;
validate_timeout(_) ->
    error(badarg).
